### PR TITLE
Rework Toolchain model and drop relative file path overrides

### LINKED ISSUE #3130: rustup allows `rustup toolchain link none`
### Problem

This is not *necessarily* a bug, but it seems very confusing.

```
; rustup default none
info: default toolchain unset
; rustup toolchain link none build/aarch64-unknown-linux-gnu/stage0
; cargo
error: rustup could not choose a version of cargo to run, because one wasn't specified explicitly, and no default is configured.
help: run 'rustup default stable' to download the latest stable release of Rust and set it as your default toolchain.
; cargo +none --version
cargo 1.67.0-beta.3 (f6e737b1e 2022-12-02)
```

### Steps

1. Clone rust-lang/rust and run `./x.py --help
2. Run `rustup toolchain link none build/aarch64-unknown-linux-gnu/stage0` (as appropriate for your host triple)
3. Run `cargo +none --version`

### Possible Solution(s)

I think disallowing `none` for a custom toolchain name would be easier to understand, something like the existing error for reserved channel names:
```
; rustup toolchain link stable build/aarch64-unknown-linux-gnu/stage0
error: invalid custom toolchain name: 'stable-aarch64-unknown-linux-gnu'
```

### Notes

_No response_

### Rustup version

```console
rustup 1.25.1 (bb60b1e89 2022-07-12)
```

### Installed toolchains

```console
> rustup show -v
Default host: aarch64-unknown-linux-gnu
rustup home:  /home/gh-jyn514/.local/lib/rustup

installed toolchains
--------------------

error: no default toolchain configured

```
(this seems like a second bug: I definitely have at least nightly installed.)

### PR DESCRIPTION
**Pull request by rbtcollins**
Rework Toolchain model and drop relative file path overrides
    
Relative path overrides permit a freshly downloaded source tree to
execute arbitrary code on any rustup command that executes a binary from
the configured toolchain, and its a reasonable tradeoff for us to remove
this feature. Absolute path overrides are kept intact - these were added
to support users of large monorepo tool systems, and can be kept with
reasonable safety.
    
Introduce an interior model for toolchain names which allows the
elimination of conditional code for custom vs named vs path based
toolchains. Fixes #3130. Close https://github.com/rust-lang/rustup/pull/3174
    
Finishes the separation of Toolchain vs DistributableToolchain vs
CustomToolchain started some time ago.
    
Moves some test support into the src tree in order to keep symbols
private and adds a 'test' feature required for self-testing, ensuring
that that code doesn't affect production builds.

Adds in a new build that tests the combination of features together (on just one platform), to ensure all combinations are valid.

### CODE REVIEW

File: src/cli/common.rs [hunk 1]
```diff
@@ -178,51 +181,71 @@ pub(crate) fn set_globals(verbose: bool, quiet: bool) -> Result<Cfg> {
 
 pub(crate) fn show_channel_update(
     cfg: &Cfg,
-    name: &str,
+    name: PackageUpdate,
```
Comment (0xPoe) [inline src/cli/common.rs:184]: Perhaps we should consider using a different term than "name"?

File: src/toolchain/names.rs [hunk 2]
```diff
@@ -0,0 +1,712 @@
+//! Overview of toolchain modeling.
+//!
+//! From the user (including config files, toolchain files and manifests) we get
+//! a String. Strings are convertable into `MaybeOfficialToolchainName`,
```
Comment (0xPoe) [inline src/toolchain/names.rs]:
```suggestion
//! a String. Strings are convertible into `MaybeOfficialToolchainName`,
```

File: tests/suite/cli-ui/rustup/rustup_unknown_arg_stdout.toml [hunk 3]
```diff
@@ -3,7 +3,7 @@ args = ["random"]
 status.code = 1
 stdout = ""
 stderr = """
-error: Invalid value \"random\" for '<+toolchain>': \"random\" is not a valid subcommand, so it was interpreted as a toolchain name, but it is also invalid. To override the toolchain using the 'rustup +toolchain' syntax, make sure to prefix the toolchain override with a '+'
+error: Invalid value \"random\" for '<+toolchain>': error: \"random\" is not a valid subcommand, so it was interpreted as a toolchain name, but it is also invalid. To override the toolchain using the 'rustup +toolchain' syntax, make sure to prefix the toolchain override with a '+'
```
Comment (0xPoe) [inline tests/suite/cli-ui/rustup/rustup_unknown_arg_stdout.toml:6]: There are two `error:`. It is a little bit confusing.

File: src/toolchain/names.rs [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+    //         let candidate = validate(candidate)?;
+    //         if candidate.parse::<PartialToolchainDesc>().is_ok()
+    //             || candidate == "none"
+    //             || candidate.contains('/')
+    //             || candidate.contains('\\')
+    //         {
+    //             Err(InvalidName::CustomName(candidate.into()))
+    //         } else {
+    //             Ok(CustomToolchainName(candidate.into()))
+    //         }
```
Comment (0xPoe) [inline src/toolchain/names.rs]: Can we remove this?

**Review by 0xPoe — APPROVED** [review]
I have noticed a few minor issues. 

Looks great and much clearer. Thank you!

File: src/cli/common.rs (hunk 1 above)
Comment (rbtcollins) [reply src/cli/common.rs:184]: I think a future cleanup can just remove this intermediate function entirely

File: src/toolchain/names.rs (hunk 4 above)
Comment (rbtcollins) [reply src/toolchain/names.rs]: oh, yes, I had it here for reference.

**Comment by matklad** [comment]
>Relative path overrides permit a freshly downloaded source tree to
execute arbitrary code on any rustup command that executes a binary from
the configured toolchain, and its a reasonable tradeoff for us to remove
this feature. Absolute path overrides are kept intact - these were added
to support users of large monorepo tool systems, and can be kept with
reasonable safety.

I think relative paths are a subset of absolute paths, so, security-wise, this is a no-op.

Namely, `./foo/bar` relative path is equivalent to `/proc/self/cwd/foo/bar` absolute path.

**Comment by matklad** [comment]
Verified with

```
[toolchain]
# path = ".cargo"
path = "/proc/self/cwd/.cargo"
```

on

https://github.com/jonas-schievink/mallory

**Comment by jonhoo** [comment]
That's definitely unfortunate. I think removing path-based toolchains altogether is going to break things, especially for more corporate users with additional build orchestration around Cargo (e.g., Amazon). How feasible is it to disallow specifically `/proc` (and maybe also `/sys`)? Do we know of other absolute-to-relative proxies like this?

**Comment by matklad** [comment]
@rbtcollins is the issue that absolute paths are also unsafe tracked anywhere?

**Comment by rbtcollins** [comment]
Not yet, please do open a bug from this, and we'll have to discuss whether to give up, filter some absolute paths and hope that's sufficient, or some other thing.

**Pull request merged**